### PR TITLE
Add AddGaussianNoise Transform

### PR DESCRIPTION
**Pull request by amurthy1**
Adds a transform for adding Gaussian Noise to a Tensor. Useful to apply when training image models to reduce over fitting.

Tested using unit tests. Wasn't sure how to test the docs exactly, open to suggestions on that.

### CODE REVIEW

File: torchvision/transforms/transforms.py [hunk 1]
```diff
@@ -1375,3 +1375,47 @@ def __call__(self, img):
...
+    Args:
+        mean (float): mean of the Gaussian distribution from which to sample.
+        std (float): std of the Gaussian distribution from which to sample.
+        p (float): probability that noise is added.
+
+    Returns:
+        Tensor with Gaussian noise added.
+    """
+
+    def __init__(self, mean=0., std=1., p=1.):
```
Comment (bmanga) [inline torchvision/transforms/transforms.py:1396]: I know that the parameter order is all over the place across these classes, but would it make sense for p to be the first one?
Also, maybe a default of 0.5 would be more consistent with the other transforms.

**Comment by amurthy1** [comment]
Can someone help review this?

**Comment by fmassa** [comment]
Hi,

Thanks for the PR!

Would you have references in research papers describing or mentioning using this transformation, or showing that it is helpful compared to other color augmentations? It would be good to have references to at least one paper in the documentation so that users can refer to it when considering using this transformation.

**Comment by amurthy1** [comment]
Hi @fmassa, thanks for the review!

I came across this transform when working with [this paper](https://arxiv.org/pdf/1711.04340.pdf) which uses the transform as I've written.

I thought this was a fairly standard transform, as it is in [tensorflow](https://www.tensorflow.org/api_docs/python/tf/keras/layers/GaussianNoise) and there have been multiple requests for it in torch ([here](https://discuss.pytorch.org/t/how-to-add-noise-to-mnist-dataset-when-using-pytorch/59745) and [here](https://github.com/pytorch/vision/issues/712))

Upon further research at your request, it seems this is more popular for traditional neural networks ([this link has several textbook references](https://machinelearningmastery.com/train-neural-networks-with-noise-to-reduce-overfitting/#:~:text=In%20effect%2C%20adding%20noise%20expands,simple%20form%20of%20data%20augmentation)) and less proven for the vision use case. 

[This](https://europepmc.org/article/pmc/pmc5977656) was the only non-paywall paper I could find that did a direct comparison of augmentation techniques and showed that noise performed worse than other augmentation techniques, albeit limited to a specific vision use case.

It seems inconclusive whether this is actually an impactful transform for vision, but it seems there is demand for it from torch users, so still might be worth adding. Let me know how you want to proceed.

**Comment by fmassa** [comment]
Hi @amurthy1 

I had a look at the DAGAN paper, and I have the impression that they used a slight different setup in their use-case: they take a random vector and pass it together with the encoded image representation, so that they augment the number of training data after feeding it to the decoder 
image
which seems to be different than what we are doing here.

## GaussianNoise in TensorFlow

I've looked a bit further and the reference that you point out for `GaussianNoise` in TensorFlow came from Keras in https://github.com/keras-team/keras/pull/335, where they referenced to "Stacked denoising autoencoders: Learning useful representations in a deep network with a local denoising criterion"  for learning auto-encoders back in 2010.

## More references for GaussianNoise

I've searched a bit more as well in the self-supervised domain, where people add different data augmentation for different branches of the model for it to learn powerful features in an non-supervised way, as it would be a good place to get some more signal.
Looking for example at SimCLR, even though I found [references of additive GaussianNoise in their paper](https://arxiv.org/pdf/2002.05709.pdf) as follows
image
their models and [official implementation do not use it](https://github.com/google-research/simclr/blob/master/data_util.py) but only random color augmentation. Given that they do mention random noise and they don't use it, I'm tempted to assume that it doesn't actually help.

Overall, I think that my previous comment from https://github.com/pytorch/vision/issues/712#issuecomment-452691229 still applies, as this function can be implemented with a one-liner via `LambdaTransform` and given that there has not really been any conclusive evidence that this transform is helpful in recent research papers, I'm tempted to say that it is better not to merge this.

As such, I'm closing this PR, but thanks again for sending the PR, and for sparkling this discussion. Let me know if you think otherwise or disagree.

**Comment by amurthy1** [comment]
I agree this is probably the right decision to not merge this given the new information. Just want to add for posterity that this is the relevant part of the DAGAN paper which I think aligns with the PR

Screenshot 2020-07-06 14 27 32

Regardless, I appreciate you taking the time to dig into this.

**Comment by fmassa** [comment]
Thanks for pointing out the part that I missed in the DAGAN paper. Let us know if you would like to contribute other features to torchvision, there are a number of issues that would be great to be addressed, for example https://github.com/pytorch/vision/issues?q=is%3Aopen+is%3Aissue+label%3A%22high+priority%22
